src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,648 @@
+        std::vector<Literal> notsatisfied = current;
+        HGenerate(notsatisfied, l, rules, graph, pattern, embeddings, forbidden_rules, sigma);
+    }
+    return rules;
+}
+
+void AddCompacted(std::vector<SingleRule>& rules, std::vector<Gfd>& result, graph_t& pattern,
+                  std::vector<Rules>::iterator& forbidden_rules) {
+    if (!rules.empty()) {
+        std::sort(rules.begin(), rules.end());
+        std::vector<Literal> prev = rules.front().first;
+        std::vector<Literal> conclusion = {};
+        for (auto const& [premises, l] : rules) {
+            if (prev != premises) {
+                Gfd gfd = {pattern, prev, conclusion};
+                result.push_back(std::move(gfd));
+                forbidden_rules->push_back({prev, conclusion});
+                prev = premises;
+                conclusion = {l};
+            } else {
+                conclusion.push_back(l);
+            }
+        }
+        Gfd gfd = {pattern, prev, conclusion};
+        result.push_back(std::move(gfd));
+        forbidden_rules->push_back({std::move(prev), std::move(conclusion)});
+    }
+}
+
+void HorizontalSpawn(graph_t const& graph, std::size_t const sigma,
+                     std::vector<graph_t> const& patterns,
+                     std::vector<Embeddings> const& embeddings_set,
+                     std::vector<Rules>& forbidden_rules_set,
+                     std::map<std::string, Info> const& attrs_info, std::vector<Gfd>& result) {
+    for (std::size_t i = 0; i < patterns.size(); ++i) {
+        auto pattern = patterns.at(i);
+        auto embeddings = embeddings_set.at(i);
+        auto forbidden_rules = std::next(forbidden_rules_set.begin(), i);
+
+        std::vector<Literal> literals = GenerateLiterals(pattern, attrs_info);
+        std::vector<SingleRule> rules =
+                GenerateRules(literals, graph, pattern, embeddings, forbidden_rules, sigma);
+        AddCompacted(rules, result, pattern, forbidden_rules);
+    }
+}
+
+class CompareCallback {
+private:
+    graph_t const& query_;
+    Embedding& iso_;
+    bool& res_;
+
+public:
+    CompareCallback(graph_t const& query, Embedding& iso, bool& res)
+        : query_(query), iso_(iso), res_(res) {}
+
+    template <typename CorrespondenceMap1To2, typename CorrespondenceMap2To1>
+    bool operator()(CorrespondenceMap1To2 f, CorrespondenceMap2To1) const {
+        BGL_FORALL_VERTICES_T(u, query_, graph_t) {
+            iso_.emplace(u, get(f, u));
+        }
+        res_ = true;
+        return false;
+    }
+};
+
+struct VCompare {
+    graph_t const& query;
+    graph_t const& graph;
+
+    bool operator()(vertex_t const& fr, vertex_t const& to) const {
+        return query[fr].attributes.at("label") == graph[to].attributes.at("label");
+    }
+};
+
+struct ECompare {
+    graph_t const& query;
+    graph_t const& graph;
+
+    bool operator()(edge_t const& fr, edge_t const& to) const {
+        return query[fr].label == graph[to].label;
+    }
+};
```

Comment:
Can we use lambdas instead of these structs?
Define them directly where you need them.

Reply:
Unfortunately, the boost graph library requires exactly this structure for the code to work correctly.